packages/node_modules/pouchdb-adapter-indexeddb/src/find.js

```diff
@@ -0,0 +1,256 @@
+'use strict';
+
+import {
+  DOC_STORE,
+  idbError,
+  rawIndexFields,
+  naturalIndexName
+} from './util';
+
+import {
+  IDB_NULL,
+  IDB_TRUE,
+  IDB_FALSE,
+} from './rewrite';
+
+// Adapted from
+// https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-find/src/adapters/local/find/query-planner.js#L20-L24
+// This could change / improve in the future?
+var COUCH_COLLATE_LO = null;
+var COUCH_COLLATE_HI = '\uffff'; // actually used as {"\uffff": {}}
+
+// Adapted from: https://www.w3.org/TR/IndexedDB/#compare-two-keys
+// Importantly, *there is no upper bound possible* in idb. The ideal data
+// structure an infintely deep array:
+//   var IDB_COLLATE_HI = []; IDB_COLLATE_HI.push(IDB_COLLATE_HI)
+// But IDBKeyRange is not a fan of shenanigans, so I've just gone with 12 layers
+// because it looks nice and surely that's enough!
+var IDB_COLLATE_LO = Number.NEGATIVE_INFINITY;
+var IDB_COLLATE_HI = [[[[[[[[[[[[]]]]]]]]]]]];
+
+//
+// TODO: this should be made offical somewhere and used by AllDocs / get /
+// changes etc as well.
+//
+function externaliseRecord(idbDoc) {
+  var doc = idbDoc.revs[idbDoc.rev].data;
+  doc._id = idbDoc.id;
+  doc._rev = idbDoc.rev;
+  if (idbDoc.deleted) {
+    doc._deleted = true;
+  }
+
+  return doc;
+}
+
+/**
+ * Generates a keyrange based on the opts passed to query
+ *
+ * The first key is always 0, as that's how we're filtering out deleted entries.
+ */
+function generateKeyRange(opts) {
+  function defined(obj, k) {
+    return obj[k] !== void 0;
+  }
+  function fix(key, exact) {
+    var fixed = [0].concat(key);
+
+    // REVIEWER: Read this block carefully.
+    //
+    // This code is replacing *all* instances both the upper and lower collate
+    // mark, regardless of where they are in the array, for non exact (eg
+    // ranged) queries.
+    //
+    // I am not sure if this is right.
+    fixed = fixed.map(function (k) {
```

Comment:
I think you can just return here.

Reply:
I don't think fix is a good name for the function. `Normalise` would be better.

Reply:
Or something related to the fact we changing the keys to work with the way we store the data.

Reply:
`translate`?

Reply:
`codify` ?

Reply:
I've gone with convert for now. I think that makes the most sense, its goal is to take a key that is valid for CouchDB and return one that is valid for IndexedDB.

We can always change it in a few if we take a later look and it's confusing.